The current state and future plan of the waku-rln-relay project (#143)
vacp2p/research: rln-research/2022-07-22-rln-relay-journey.md

---
layout: post
name:  ""
title: ""
date: 2022-10-18 10:00:00
author: sanaztaheri
published: true
permalink: /waku-rln-relay-journey
categories: research

> summary: The current state of the research and development of the WAKU-RLN-RELAY project.

Reviewer: This summary could be better:

- Doesn't help someone who doesn't already know what RLN or RLN RELAY is.

- It also doesn't hint at what the main delta is from last post (e.g. cross client on-chain testnet, etc) - can use a few keywords in a list to give people a better view of what the post about.

The first XX words of the summary shows up in e.g. Twitter view

Author: > Doesn't help someone who doesn't already know what RLN or RLN RELAY is.

Fair point.

> It also doesn't hint at what the main delta is from last post (e.g. cross client on-chain testnet, etc) - can use a few keywords in a list to give people a better view of what the post about.

Sure.

> The first XX words of the summary shows up in e.g. Twitter view

Noted.

Please check the following commit https://github.com/vacp2p/research/pull/143/commits/5e3ab814150340248272524c2473c4ca60346856

image: 
discuss: 
_includes: [math]
---

In this post, we would like to shed light on the current state of the research and development of the WAKU-RLN-RELAY project as well as its future direction and the areas of work to which the readers of this post may wish to contribute.

# WAKU-RLN-RELAY: Why and How

> **Why**: The WAKU-RLN-RELAY project was started more than two years ago to protect the open p2p messaging protocol of [WAKU-RELAY](https://rfc.vac.dev/spec/11/) against spam and DDoS attacks.

Reviewer: Link back to previous post on RLN

Author: sure, done https://github.com/vacp2p/research/pull/143/commits/33d5b37ec63e1271cf2e6a384bc0342d92fb3c30

> **Why**: The WAKU-RLN-RELAY project has been started more than two years ago to protect the open p2p messaging protocol of [WAKU-RELAY](https://rfc.vac.dev/spec/11/) against spam and DDoS attacks.

Reviewer 2: ```suggestion
**Why**: The WAKU-RLN-RELAY project was started more than two years ago to protect the open p2p messaging protocol of [WAKU-RELAY](https://rfc.vac.dev/spec/11/) against spam and DDoS attacks.
```

However, the use case of WAKU-RLN-RELAY is not limited to the WAKU-RELAY protocol and can be adopted by any open p2p routing system.
It is worth mentioning that WAKU-RELAY is a minor extension of [Libp2p Gossipsub protocol](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/README.md), hence the spam protection property of WAKU-RLN-RELAY is directly applicable to the libp2p Gossipsub protocol too. 

> In the nutshell, spam protection is achieved by limiting the messaging rate of every network participant so that no user can flood the network with a large number of messages and congest/DDoS relayers.

Reviewer 2: ```suggestion
In a nutshell, spam protection is achieved by limiting the messaging rate of every network participant so that no user can flood the network with a large number of messages and congest/DDoS relayers.
```

One special focus of WAKU-RLN-RELAY is user privacy which is aligned with the WAKU-RELAY privacy objectives.
That is, it does not rely on any personally identifiable information of users like IP address or peer ID to control their messaging rate.  

> This makes it superior in terms of privacy guarantees compared to the centralized system where users are asked to share some PII like phone number, or email address for network participation.

Reviewer 2: ```suggestion
This makes it superior in terms of privacy guarantees compared to centralized systems where users are asked to share some PII like a phone number, or an email address for network participation.
```

Moreover, WAKU-RLN-RELAY is superior to its privacy-preserving counterparts i.e., proof-of-work in terms of performance.

> The computational overhead imposed by proof-of-work is not suitable for resource-limited devices like mobile phones whereas the computational overhead incurred by the WAKU-RLN-RELAY protocol is minimal and as of the time of this post is approximately less than half a second. 

Reviewer 2: ```suggestion
The computational overhead imposed by proof-of-work is not suitable for resource-limited devices like mobile phones whereas the computational overhead incurred by the WAKU-RLN-RELAY protocol is minimal and at the time of writing this post is approximately less than half a second. 
```

**How**: 
Network participants which want to send messages in the network should be registered in the RLN group.
The state of the group is moderated by a smart contract which holds a list of enrolled members.
Each participant has a secret identity key and registers a commitment of that to the membership smart contract and deposits some funds.

> The key to withdraw the locked fund is the user's secret identity key, so if someone knows someone's secret identity key then it can take the associated fund.

Reviewer 2: ```suggestion
The key to withdraw the locked funds is the user's secret identity key, so if someone knows someone else's secret identity key, they can withdraw the associated funds.
```

Messaging rate is enforced per some time interval called an epoch.
For each epoch, the user sends a message by proving that it knows the opening to one of the commitments registered in the RLN group as well as it reveals a Shamir share of its secret id key. 
The share is derived in such a way that two of such shares for the same epoch allow the construction of the corresponding secret id key. 
The secret sharing has another element which is an application-defined identifier i.e., RLN identifier by which you can make the secret sharing unique to your application. 

> That is, if a user wants to use the same RLN membership credential for messaging in two or more different applications, then it can do so without being worried to get slashed by combining the shares of its identity key published in those applications.

Reviewer 2: ```suggestion
That is, if a user wants to use the same RLN membership credential for messaging in two or more different applications, then they can do so without being slashed by combining the shares of its identity key published in those applications.
```

Should there be a statement here on *why* users shouldn't use the same membership credentials for different applications? (anonymity guarantees)

Author: If users do not use the _same credentials_ in _different groups_, they should have no security concerns. What is pictured here is a different scenario and that is a user uses its _membership to one single rln group_ to message in _different applications_. Does that answer your question?

Reviewer 2: Ah, I see! I was of the opinion that by application it implied a different membership group.

## Current State

**Specifications**: The specification of the WAKU-RLN-RELAY protocol is currently in draft mode and available in [17/WAKU2-RLN-RELAY](https://rfc.vac.dev/spec/17/).
In a separate effort, as part of a collaboration with the PSE group, the specification of the RLN construct, as a stand-alone primitive,  has been standardized and published in [32/RLN](https://rfc.vac.dev/spec/32/).

> **Academic publications**: The results of this project are published as 2 academic research articles in prestigious conferences of ICDCS and DINPS 2022.

Author: Sure, done https://github.com/vacp2p/research/pull/143/commits/cd6791e42a20acc866589d3fb87a0e13ca95e4ea

> They are accessible at https://arxiv.org/pdf/2207.00117.pdf and https://ieeexplore.ieee.org/abstract/document/9912176/. 

Author: No, they are different, one is a poster and the other one is a full extended paper.

> They are accessible at https://arxiv.org/pdf/2207.00117.pdf and https://ieeexplore.ieee.org/abstract/document/9912176/. 

Reviewer: I think what s1fr0 is getting at is that this should be mentioned in the text. A question in a PR review of a text is usually feedback on the text itself, and just answering the PR comment doesn't actually improve the text...

**Testnets**: Two testnets are launched so far:
- Testnet1: In this testnet, we implemented the PoC of WAKU-RLN-RELAY in Nim and integrated it into a sample chat application. More on this can be found in [testnet1 tutorial](https://github.com/status-im/nwaku/blob/master/docs/tutorial/rln-chat2-live-testnet.md).
The testnet launched successfully and we got great feedback from the participants to improve upon user experience and resolved identified concerns and comments. 

> - Testnet2: This testnet aimed to integrate WAKU-RLN-RELAY into all 3 waku clients namely, Nim-waku, Go-Waku, and js-Waku and test their interoperability over the waku network. 

Reviewer 2: ```suggestion
- Testnet2: This testnet aimed to integrate WAKU-RLN-RELAY into all 3 waku clients namely, nwaku, Go-Waku, and js-Waku and test their interoperability over the waku network. 
```

> - Testnet2: This testnet aimed to integrate WAKU-RLN-RELAY into all 3 waku clients namely, Nim-waku, Go-Waku, and js-Waku and test their interoperability over the waku network. 

Author: Using `Waku` with the capital `W` was intentional. :)

Should you be interested in taking part in the testnet, or have any questions, let us know in our dedicated discord channel [WAKU-RLN-RELAY dogfooding](https://discord.com/channels/864066763682218004/945761301156753448).

**Zerokit**: The use of zero-knowledge proofs and zkSNARKs to provide privacy-enhanced service is not limited to the WAKU-RLN-RELAY protocol. 
The design and development of all the waku protocols are centered around privacy protection and anonymity. 
This made us realize the need to develop our zero-knowledge toolings to embrace this need.
The [Zerokit library](https://github.com/vacp2p/zerokit) was born out of this desire to build a self-hosted library to do research and exploration in the zkp field.
Zerokit currently comprises the RLN module and is going to be extended by other novel relevant constructs.
One of the driving factors of Zerokit is to facilitate the use of certain Zero Knowledge modules from other environments, namely, systems programming environments, command-line, and mobile environments. 
More on the Zerokit positioning is available in [Zerokit positioning](UAFvOj1tTQ6uW2li0ZeXfw).
  

> **Incentivization**: Spam protection will not be realized unless with a sufficient number of relayers choose to run WAKU-RLN-RELAY and protect the p2p messaging network against spammers.

Reviewer 2: ```suggestion
**Incentivization**: Spam protection will not be realized unless a sufficient number of relayers choose to run WAKU-RLN-RELAY and protect the p2p messaging network against spammers.
```

> Since spam activities and DDoS attacks directly target and exhaust network resources, it is conceivable that relayers would be naturally incentivized to run WAKU-RLN-RELAY to save their local resources from being drained by attackers. 

Reviewer 2: ```suggestion
Since spam activities and DDoS attacks directly target and exhaust network resources, it is conceivable that relayers would be naturally incentivized to run WAKU-RLN-RELAY to prevent their local resources from being used by attackers. 
```

Nevertheless, WAKU-RLN-RELAY comes with built-in economic incentives where 
relayers who detect spammers are rewarded with the ability to withdraw the spammer's staked funds from the membership contract.
However, not every node gets to find a spammer and be rewarded for that.
Indeed, due to the cryptographically solid economic punishment in place, fewer spam activities are expected to be observed.
This led us to introduce the service fee option to the RLN membership contract to compensate relayers for their spam-protection service (regardless of whether they spot any spamming activity or not).
That is, a portion of the membership registration fee goes towards a non-refundable service fee.
The distribution of the collected service fee is still a work in progress and its current state can be seen in the [rln contract repository](https://github.com/vacp2p/rln-contract).

**Support for ERC20 tokens as RLN membership fee**: It is conceivable that many applications desire to have their tokens and want their RLN group membership fee to be paid in that token.

> To enable so, we are working towards updating the RLN contract interface to support the ERC20 token for the membership fee. 

Reviewer 2: ```suggestion
To enable this, we are working towards updating the RLN contract interface to support the ERC20 token for the membership fee. 
```

**RLN trusted setup**: Down the line, we will conduct the MPC ceremony to generate parameters specific to the RLN circuit.

**Hybrid architecture for the membership Merkle tree storage management (super-peer/light-peer)**: In the current design of RLN-relay, every peer needs to persist and maintain the membership Merkle tree locally. 
While the storage overhead of Merkle tree is reasonable as discussed in the [WAKU-RLN-RELAY current state](#current-state), 
many applications may still wish to save that storage space in favor of other critical functionalities.
To allow this, we are looking into a hybrid architecture in which only a subset of peers namely, super peers persist and maintain the membership Merkle tree.
Other peers namely light peers fetch their necessary information from the super peers via a req/response protocol.
Note that as in any req/reply protocol, privacy should be carefully addressed as the service providers may be able to infer the position of the requester's id commitment in the membership Merkle tree from the requests patterns. 

> **Zero-entry**: Some applications may want to allow their users to benefit from a spam-protected network and publish messages without having any crypto account. 

Reviewer 2: ```suggestion
**Zero-entry**: Some applications may want to allow their users to benefit from a spam-protected network and publish messages without having any crypto assets. 
```
